qutip/core/coefficient.py

```diff
@@ -389,32 +386,39 @@ def make_cy_code(code, variables, constants, raw, compile_opt):
 from qutip.core.cy.complex_math cimport *
 from qutip.core.data cimport Data
 cdef double pi = 3.14159265358979323
-{}
+{compile_opt['extra_import']}
 
-parsed_code = "{}"
+parsed_code = "{code}"
 
 @cython.auto_pickle(True)
 cdef class StrCoefficient(Coefficient):
     cdef:
         str codeString
-{}{}
+{cdef_cte}{cdef_var}
 
     def __init__(self, base, var, cte, args):
         self.codeString = base
-{}{}        self.arguments(args)
-
-    cpdef void arguments(self, dict args) except *:
-{}        pass
+{init_cte}{init_var}{init_arg}
+
+    cpdef Coefficient copy(self):
+        cdef StrCoefficient out = StrCoefficient.__new__(StrCoefficient)
+        out.codeString = self.codeString
+{copy_cte}{copy_var}
+        return out
+
+    def replace(self, *, dict arguments=None, tlist=None):
```

Comment:
I asked a similar thing about the signature of `replace`, since quite a few coefficients take in `tlist` and ignore it.  I was wondering if it would be better to have the signature be `replace(**kwargs)` in all cases, and have each implementation ignore all kwargs it doesn't care about.  Then we could document only the needed kwargs in the docstring, but we'd still have the desirable behaviour that you could indiscriminately call `replace` on a higher-order object (say `QobjEvo`), and nothing would need to worry about what accepts what.  It's also potentially more extensible, because if we add a new coefficient type, it wouldn't be such a huge effort if it has additional parametrisation.

Reply:
I made it so the signature always include `**kwargs` and the variable that can be directly replaced. So for function coefficient, it is `replace(arguments=None, **kwargs)`.

Reply:
The docstrings should still say which keyword arguments are valid, though, so the user knows.  Otherwise there'd be no way to know what was valid.